Get ready for article (#1)
Bnaya/mst-typescript-dec-test: README.md

> # mobx-state-tree typescript slowness & duplications in `d.ts` emit (And how to make it better) [part 1]

Reviewer: ```suggestion
# mobx-state-tree TypeScript slowness & duplication in `d.ts` emit (and how to make it better) [part 1]
```

> > This article contains work, knowledge and experience accumulated by others and myself regrading improving the performance of typescript build, edit and IDE responsiveness.  

Reviewer: ```suggestion
> This article contains work, knowledge and experience accumulated by others and myself regarding improvements to TypeScript performance of builds, editing and IDE responsiveness.  
```

> You can find almost all this information scattered in github issues, comments and code snippets.  

Reviewer: ```suggestion
You can find almost all this information scattered across multiple GitHub issues, comments and code snippets.  
```

I will try to provide references, citations and bibliography. Please help me with that :)

> `mobx-state-tree` is an interesting creation that takes types to a new level and typescript inference power to it's limits.  

Reviewer: ```suggestion
`mobx-state-tree` is an interesting creation that takes types to a new level and 
pushes TypeScript inference to its limits.  
```

> It's doing so by defining runtime models and types in code, and inferring design-time typescript informatio.

Reviewer: ```suggestion
It's doing so by defining runtime models and types in code and inferring design-time TypeScript information.
```

> On this series of articles, we will learn about some caveats and how to overcome them.

Reviewer: ```suggestion
In this series of articles, we will learn about some caveats to this technique and how to overcome them.
```

## The problem(s)

> Emitting typescript types for naive `MST` code, that uses any form of compositing,

Reviewer: ```suggestion
Emitting TypeScript types for naive `MST` code, that uses any form of composition
```

> will lead to duplication/inline of the entire tree of model types on each reference point.

Reviewer: ```suggestion
will lead to duplication/inlining of the entire tree of model types on each reference point.
```

> will lead to duplication/inline of the entire tree of model types on each reference point.

Reviewer: What's do you mean by `reference point`?

Example:
```ts

> export const Store1 = t
>     .model('Store1', {
>         store2: t.optional(t.array(Store2), [])

Reviewer: ```suggestion
import { types } from 'mobx-state-tree'
export const Store1 = types
    .model('Store1', {
        store2: types.optional(types.array(Store2), [])
```

        store3: import("mobx-state-tree").IArrayType<import("mobx-state-tree").IModelType<{
            store4: import("mobx-state-tree").IModelType<{
                store5: import("mobx-state-tree").IModelType<{
                    prop1: import("mobx-state-tree").IMaybe<import("mobx-state-tree").ISimpleType<string>>;
                }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>;
            }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>;
            store5: import("mobx-state-tree").IOptionalIType<import("mobx-state-tree").IModelType<{
                prop1: import("mobx-state-tree").IMaybe<import("mobx-state-tree").ISimpleType<string>>;
            }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>, [undefined]>;
        }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>>;
    }, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>>, [undefined]>;
}, {}, import("mobx-state-tree")._NotCustomized, import("mobx-state-tree")._NotCustomized>;
```

> What we can see here is that instead of referencing types of the child models, we get full duplication/inlining of child models declarations.

Reviewer: ```suggestion
What we can see here is that instead of referencing types of the child models, we get full duplication/inlining of child model declarations.
```

> That happens for each level inside the stores tree will have the same duplication up to it's level.

Reviewer: ```suggestion
This happens at each level in the models tree and will have the same duplication up to its level.
```

> ## ...But it's not impact only types emit!

Reviewer: I don't understand this sentence

> You can think of types emit as visualization of the inferred data structures inside typescript language server. duplication in `d.ts` = more work `tsc/tsserver` need to do or regular basis.  

Reviewer: ```suggestion
You can think of types emit as visualization of the inferred data structures inside TypeScript's language server. Duplication in `d.ts` means more work for `tsc/tsserver` to do for any operation.  
```

> And indeed when working with big and nested models, the build time and IDE responsiveness of typescript types is badly degraded

Reviewer: ```suggestion
And indeed when working with big and nested models, TypeScript's build time and IDE responsiveness is severely degraded.
```

## Peeling the onion 🧅

> On the coming articles, we will cover issues with `MST` and workarounds for them, one by one, like peeling a fire roasted onion to get to the tasty parts!

Reviewer: ```suggestion
Across the upcoming articles, we will cover issues with `MST` and the workarounds for them, one by one, like peeling a fire roasted onion to get to the tasty parts!
```

> ## Layer one: reduce import() references

Reviewer: ```suggestion
## Layer one: reduce `import()` references
```

> When examining the emitted types for `Store1` above, the first thing that catches my eye is the many references to `import("mobx-state-tree")` for the same symbol. 

Reviewer: ```suggestion
When examining the emitted types for `Store1` above, the first thing that catches the eye is the multitude of references to `import("mobx-state-tree")` for the same symbol. 
```

`import("mobx-state-tree").IModelType` appears 5 times!

What can we do about it? Lets import the types directly!  

> Adding `import type * as MST from "mobx-state-tree";` to our model file will turn the emitted types into:

Reviewer: ```suggestion
Adding `import type * as MST from "mobx-state-tree";` to our model file will shorten the emitted types into:
```

```ts
import type * as MST from "mobx-state-tree";
export declare const Store1: MST.IModelType<{
    store2: MST.IOptionalIType<MST.IArrayType<MST.IModelType<{
        store3: MST.IArrayType<MST.IModelType<{
            store4: MST.IModelType<{
                store5: MST.IModelType<{
                    prop1: MST.IMaybe<MST.ISimpleType<string>>;
                }, {}, MST._NotCustomized, MST._NotCustomized>;
            }, {}, MST._NotCustomized, MST._NotCustomized>;
            store5: MST.IOptionalIType<MST.IModelType<{
                prop1: MST.IMaybe<MST.ISimpleType<string>>;
            }, {}, MST._NotCustomized, MST._NotCustomized>, [undefined]>;
        }, {}, MST._NotCustomized, MST._NotCustomized>>;
    }, {}, MST._NotCustomized, MST._NotCustomized>>, [undefined]>;
}, {}, MST._NotCustomized, MST._NotCustomized>;

```

Much more compact and readable!  

> You may also import individual symbols instead of a star import

Reviewer: ```suggestion
You may also import individual symbols instead of a star import.
```

## Layer two: Child stores inlining

> The second thing that catches the eye is that, even thu `Store2` (And the rest) are declared on a separate file, we ended up with the entire stores types inline inside `Store1`.

Reviewer: ```suggestion
The second thing that catches the eye is that, even though `Store2` (as well as the rest of the stores) are declared in a separate file, they're still inlined inside `Store1`.
```

Let's understand why that's happing:

> When typescript does not have a clear name for type information, it will be inlined into his reference points!  

Reviewer: ```suggestion
When TypeScript does not have a clear name for type information, it will be inlined into its reference points!  
```

Reviewer: What are `reference points` exactly? If that's a thing which you didn't invent, add a link as a reference.

> MST `models` and `types` are not TypeScript type information, but actual javascript values, that typescript infers type information from, and creates ephemeral interfaces/types without a name.

Reviewer: ```suggestion
MST `model`s and `types` are not TypeScript types, but actual JavaScript values, from which TypeScript infers type information, creating ephemeral and anonymous interfaces/types.
```

Reviewer: Where's the term "ephemeral types/interfaces" from? Add a link to it.

> ### How can we name inferred, "ephemeral" types?

Reviewer: ```suggestion
### How can we name inferred, ephemeral types?
```

> We will create an interface out of inferred type information, and mask the original model with it!

Reviewer: ```suggestion
We will create an interface from the inferred type information, masking the original model with it!
```

> Before, "bad" example:

Reviewer: ```suggestion
Before, "bad" example input:
```

{% gist https://gist.github.com/Bnaya/5c208956a3e5f3699595f9b9ca0ad3eb file=bad.ts %}

Will emit:
{% gist https://gist.github.com/Bnaya/5c208956a3e5f3699595f9b9ca0ad3eb file=bad.d.ts %}

> After, good example:

Reviewer: ```suggestion
After, good example input:
```

{% gist https://gist.github.com/Bnaya/5c208956a3e5f3699595f9b9ca0ad3eb file=good.ts %}

Will emit:
{% gist https://gist.github.com/Bnaya/5c208956a3e5f3699595f9b9ca0ad3eb file=good.d.ts %}

> Examining Model B types,

Reviewer: ```suggestion
Examining `ModelB` types,
```

```ts
declare const ModelBInferred: IModelType<{
    /**
     * No Duplication here, but a reference to ModelAFactoryType
     */
    a: ModelAFactoryType;
}, {}, _NotCustomized, _NotCustomized>;
```

> We can see that Model A is no longer inlined, but the there's a reference to the interface we've created.

Reviewer: ```suggestion
Note how `ModelA` is no longer inlined, but the there's a reference to the interface we've created instead.
```

## Recap, downsides and tips 

> This way of writing MST models is very verbose, and add additional indirections, which is not fun.

Reviewer: ```suggestion
This way of writing MST models is very verbose, and adds additional indirection, which is not fun.
```

> You may selectively apply that on some of your models.  

Reviewer: ```suggestion
You may selectively apply this technique for some of your models.  
```

> It's highly recommended to have a clear naming convention for the inferred symbols, and the public symbols.  

Reviewer: ```suggestion
It's highly recommended to have a clear naming convention for the inferred symbols, and the public symbols.  
```
```suggestion
It's highly recommended to have a clear naming convention for the inferred symbols, as well as the public ones.  
```

> These patterns are not applicable only for models, but for every MST complex type. 

Reviewer: ```suggestion
This technique isn't only applicable to `model`s, but to any MST complex type. 
```

### Thank you for reading, your feedback is very welcome!

## Coming up next in part 2:

> On part 2 we will analyze why MST union still cause duplications, and how we can overcome them.  

Reviewer: ```suggestion
In part 2 we will analyze why MST's union type still causes duplication, and how we can overcome that.  
```
